mockwebserver/src/main/kotlin/mockwebserver3/internal/duplex/MockSocketHandler.kt

```diff
@@ -20,32 +20,46 @@ import java.util.concurrent.CountDownLatch
 import java.util.concurrent.FutureTask
 import java.util.concurrent.LinkedBlockingQueue
 import java.util.concurrent.TimeUnit
-import mockwebserver3.Stream
-import mockwebserver3.StreamHandler
+import mockwebserver3.SocketHandler
+import okio.BufferedSink
+import okio.BufferedSource
+import okio.Socket
+import okio.buffer
 import okio.utf8Size
 
-private typealias Action = (Stream) -> Unit
+private typealias Action = (BufferedSocket) -> Unit
+
+private class BufferedSocket(
+  val socket: Socket,
+) {
+  val source: BufferedSource = socket.source.buffer()
+  val sink: BufferedSink = socket.sink.buffer()
+
+  fun cancel() {
+    socket.cancel()
+  }
+}
```

Comment:
This little helper is a bit disappointing to need! In theory we could do `Socket` and `BufferedSocket` in Okio, as we do with `Source` and `BufferedSource`. I don’t need it in the public API yet, so I’m going to hold off.

Reply:
I think you'll definitely need it, but maybe reason to learn more before committing to it.